### PR TITLE
Implement RightToObtainData to all the Identity Events.

### PR DESCRIPTION
**Pull request by VadimSchmitz**
Gather relevant user information from EventStream, The sensitive data and non-sensitive data are combined and then held against a whitelist. The whitelisted data is returned in a JSON serializable object.

TravisCI currently fails due to a security vulnerability. this will be fixed in a future pr.
Some of the documentation is also outdated or needs some tweaking/adding. this will also be done in a future pr.

Task 3 of [#181430595](https://www.pivotaltracker.com/story/show/181430595)

### CODE REVIEW

File: src/Surfnet/Stepup/Identity/Event/CompliedWithRevocationEvent.php [hunk 1]
```diff
@@ -26,10 +26,19 @@
...
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\RightToObtainDataInterface;
 use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\SensitiveData;
 
-abstract class CompliedWithRevocationEvent extends IdentityEvent implements Forgettable
+abstract class CompliedWithRevocationEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/CompliedWithRevocationEvent.php:37]: Did you leave out the SecondFactorIdentifier by choice?

File: src/Surfnet/Stepup/Identity/Event/EmailVerifiedEvent.php [hunk 2]
```diff
@@ -30,10 +30,20 @@
...
+class EmailVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+        'registration_requested_at',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/EmailVerifiedEvent.php:45]: Why did the commonName, and email not end up on the whitelist?

File: src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php [hunk 3]
```diff
@@ -30,10 +30,20 @@
...
+class GssfPossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'stepup_provider',
+        'registration_requested_at',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php:45]: Why are the: SF Identifier, Common name and Email not on the whitelist?

File: src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenEvent.php [hunk 4]
```diff
@@ -30,10 +30,19 @@
...
-class GssfPossessionProvenEvent extends IdentityEvent implements Forgettable
+class GssfPossessionProvenEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'stepup_provider',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenEvent.php:44]: Probabliy missing:
- email verification required
- commonname
- email

File: src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaForInstitutionEvent.php [hunk 5]
```diff
@@ -118,4 +129,9 @@ public function serialize(): array
             'ra_institution'              => (string) $this->raInstitution,
         ];
     }
+
+    public function obtainUserData(): array
+    {
+        // TODO: Implement obtainUserData() method.
+    }
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaForInstitutionEvent.php:136]: Forgot to implement?

File: src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php [hunk 6]
```diff
@@ -28,8 +29,15 @@
...
 
-class IdentityCreatedEvent extends IdentityEvent implements Forgettable
+class IdentityCreatedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'id',
+        'institution',
+        'name_id',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php:39]: CommonName, Email missing?

File: src/Surfnet/Stepup/Identity/Event/IdentityEmailChangedEvent.php [hunk 7]
```diff
@@ -23,10 +23,16 @@
...
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\RightToObtainDataInterface;
 use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\SensitiveData;
 
-class IdentityEmailChangedEvent extends IdentityEvent implements Forgettable
+class IdentityEmailChangedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'id',
+        'identity_institution',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/IdentityEmailChangedEvent.php:34]: Email missing?

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenAndVerifiedEvent.php [hunk 8]
```diff
@@ -29,10 +29,19 @@
...
-class PhonePossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified
+class PhonePossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'registration_requested_at',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenAndVerifiedEvent.php:43]: Missing: SecondFactorId  PhoneNumber   CommonName  Email

File: src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityInformationAmendedEvent.php [hunk 9]
```diff
@@ -24,12 +24,21 @@
...
-class RegistrationAuthorityInformationAmendedEvent extends IdentityEvent
+class RegistrationAuthorityInformationAmendedEvent extends IdentityEvent implements RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'name_id',
+        'institution',
+        'location',
+        'contact_information',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityInformationAmendedEvent.php:40]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedForInstitutionEvent.php [hunk 10]
```diff
@@ -25,10 +25,18 @@
...
 
-class RegistrationAuthorityRetractedForInstitutionEvent extends IdentityEvent implements Forgettable
+class RegistrationAuthorityRetractedForInstitutionEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'name_id',
+        'ra_institution',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedForInstitutionEvent.php:38]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php [hunk 11]
```diff
@@ -32,13 +32,23 @@
...
+class SecondFactorVettedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'name_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php:50]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenAndVerifiedEvent.php [hunk 12]
```diff
@@ -29,10 +29,19 @@
...
-class YubikeyPossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified
+class YubikeyPossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'registration_requested_at',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenAndVerifiedEvent.php:43]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenEvent.php [hunk 13]
```diff
@@ -29,10 +29,18 @@
...
 
-class YubikeyPossessionProvenEvent extends IdentityEvent implements Forgettable
+class YubikeyPossessionProvenEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenEvent.php:42]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/YubikeySecondFactorBootstrappedEvent.php [hunk 14]
```diff
@@ -29,10 +29,19 @@
...
-final class YubikeySecondFactorBootstrappedEvent extends IdentityEvent implements Forgettable
+final class YubikeySecondFactorBootstrappedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'name_id',
+        'identity_institution',
+        'preferred_locale',
+        'second_factor_id',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/YubikeySecondFactorBootstrappedEvent.php:43]: Sensitive data is missing

File: src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php [hunk 15]
```diff
@@ -112,4 +120,11 @@ public function setSensitiveData(SensitiveData $sensitiveData)
         $this->email      = $sensitiveData->getEmail();
         $this->commonName = $sensitiveData->getCommonName();
     }
+
+    public function obtainUserData(): array
+    {
+        $serializedPublicUserData = $this->serialize();
+        $serializedSensitiveUserData = $this->getSensitiveData()->serialize();
+        return array_merge($serializedPublicUserData, $serializedSensitiveUserData);
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php:128]: Please apply the whitelist here

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php [hunk 16]
```diff
@@ -173,4 +181,11 @@ public function setSensitiveData(SensitiveData $sensitiveData)
         $this->commonName  = $sensitiveData->getCommonName();
         $this->phoneNumber = $sensitiveData->getSecondFactorIdentifier();
     }
+
+    public function obtainUserData(): array
+    {
+        $serializedPublicUserData = $this->serialize();
+        $serializedSensitiveUserData = $this->getSensitiveData()->serialize();
+        return array_merge($serializedPublicUserData, $serializedSensitiveUserData);
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php:189]: Whitelist not is not applied

File: symfony.lock [hunk 17]
```diff
@@ -50,6 +50,9 @@
     "doctrine/dbal": {
         "version": "2.10.2"
     },
+    "doctrine/deprecations": {
```
Comment (MKodde) [inline symfony.lock:53]: The change to the symfony.lock seems unrelated to this PR. Was there a good reason to add it? Otherwise please remove it from history ;)

File: src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php [hunk 18]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+/**
+ * Copyright 2016 SURFnet B.V.
```
Comment (MKodde) [inline src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php:4]: Nitpick: user current year (also goes for other new files added in this PR)

File: src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+use Surfnet\Stepup\Identity\Value\Institution;
+use Surfnet\Stepup\Identity\Value\Locale;
+use Surfnet\Stepup\Identity\Value\NameId;
+
+
+class AppointedAsRaaForInstitutionEventTest extends TestCase
+{
+    public function test_if_all_whitelisted_data_is_returned()
+    {
+        $AppointedAsRaaForInstitutionEvent = new AppointedAsRaaForInstitutionEvent(
```
Comment (MKodde) [inline src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php:36]: Convention: variables are always in lowerCamelCase

File: src/Surfnet/Stepup/Tests/Configuration/Event/IdentityCreatedEventTest.php [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+            new Locale("nl_NL"));
+
+        $userData = $IdentityCreatedEvent->obtainUserData();
+
+        $this->assertArrayHasKey("id", $userData);
+        $this->assertArrayHasKey("institution", $userData);
+        $this->assertArrayHasKey("name_id", $userData);
+        $this->assertArrayHasKey("preferred_locale", $userData);
+        $this->assertArrayHasKey("common_name", $userData);
+        $this->assertArrayHasKey("email", $userData);
```
Comment (MKodde) [inline src/Surfnet/Stepup/Tests/Configuration/Event/IdentityCreatedEventTest.php:50]: You could also test that data that is explicitly removed from the results as per whitelist filtering is not in the resultsset.

**Review by MKodde — CHANGES_REQUESTED** [review]

File: src/Surfnet/Stepup/Identity/Event/CompliedWithRevocationEvent.php (hunk 1 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/CompliedWithRevocationEvent.php:37]: added

File: src/Surfnet/Stepup/Identity/Event/EmailVerifiedEvent.php (hunk 2 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/EmailVerifiedEvent.php:45]: added

File: src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php (hunk 3 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php:45]: added  'gssfId', 'commonName', 'email',

File: src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenEvent.php (hunk 4 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenEvent.php:44]: email verification is not a whitelisted attribute,
added commonName and email

File: src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaForInstitutionEvent.php (hunk 5 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaForInstitutionEvent.php:136]: implemented

File: src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php (hunk 6 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php:39]: added

File: src/Surfnet/Stepup/Identity/Event/IdentityEmailChangedEvent.php (hunk 7 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/IdentityEmailChangedEvent.php:34]: added

File: src/Surfnet/Stepup/Identity/Event/IdentityRenamedEvent.php [hunk 21]
```diff
@@ -23,10 +23,16 @@
...
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\RightToObtainDataInterface;
 use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\SensitiveData;
 
-class IdentityRenamedEvent extends IdentityEvent implements Forgettable
+class IdentityRenamedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'id',
+        'institution',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/IdentityRenamedEvent.php:34]: added

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenAndVerifiedEvent.php (hunk 8 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenAndVerifiedEvent.php:43]: added

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php (hunk 16 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php:189]: changed implementation

File: src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityInformationAmendedEvent.php (hunk 9 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityInformationAmendedEvent.php:40]: there is no sensitive data handled in this class.

File: src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedForInstitutionEvent.php (hunk 10 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedForInstitutionEvent.php:38]: added

File: src/Surfnet/Stepup/Identity/Event/SecondFactorMigratedToEvent.php [hunk 22]
```diff
@@ -26,13 +26,22 @@
...
+class SecondFactorMigratedToEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'target_institution',
+        'target_second_factor_id',
+        'second_factor_type',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/SecondFactorMigratedToEvent.php:44]: added secondFactorIdentifier.

File: src/Surfnet/Stepup/Identity/Event/SecondFactorRevokedEvent.php [hunk 23]
```diff
@@ -26,10 +26,18 @@
...
 
-abstract class SecondFactorRevokedEvent extends IdentityEvent implements Forgettable
+abstract class SecondFactorRevokedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/SecondFactorRevokedEvent.php:39]: added

File: src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php (hunk 11 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php:50]: added

File: src/Surfnet/Stepup/Identity/Event/SecondFactorVettedWithoutTokenProofOfPossession.php [hunk 24]
```diff
@@ -32,13 +32,23 @@
...
+class SecondFactorVettedWithoutTokenProofOfPossession extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'name_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+        'preferred_locale',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/SecondFactorVettedWithoutTokenProofOfPossession.php:50]: added

File: src/Surfnet/Stepup/Identity/Event/U2fDevicePossessionProvenAndVerifiedEvent.php [hunk 25]
```diff
@@ -29,13 +29,22 @@
...
-class U2fDevicePossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified
+class U2fDevicePossessionProvenAndVerifiedEvent extends IdentityEvent implements Forgettable, PossessionProvenAndVerified, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'registration_requested_at',
+        'preferred_locale',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/U2fDevicePossessionProvenAndVerifiedEvent.php:46]: added

File: src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php (hunk 3 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/GssfPossessionProvenAndVerifiedEvent.php:45]: added

File: src/Surfnet/Stepup/Identity/Event/U2fDevicePossessionProvenEvent.php [hunk 26]
```diff
@@ -29,13 +29,21 @@
...
  */
-class U2fDevicePossessionProvenEvent extends IdentityEvent implements Forgettable
+class U2fDevicePossessionProvenEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'second_factor_id',
+        'preferred_locale',
+    ];
```
[the remark this answers is not in the document]
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/U2fDevicePossessionProvenEvent.php:45]: added

File: src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenAndVerifiedEvent.php (hunk 12 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenAndVerifiedEvent.php:43]: added

File: src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenEvent.php (hunk 13 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/YubikeyPossessionProvenEvent.php:42]: added

File: src/Surfnet/Stepup/Identity/Event/YubikeySecondFactorBootstrappedEvent.php (hunk 14 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/YubikeySecondFactorBootstrappedEvent.php:43]: added

File: src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php (hunk 15 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/IdentityCreatedEvent.php:128]: done

File: src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php (hunk 18 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php:4]: changed

File: src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php (hunk 19 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Tests/Configuration/Event/AppointedAsRaaForInstitutionEventTest.php:36]: changed

File: src/Surfnet/Stepup/Tests/Configuration/Event/IdentityCreatedEventTest.php (hunk 20 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Tests/Configuration/Event/IdentityCreatedEventTest.php:50]: added this to PhonePosessionProvenAndVerifiedEventTest since in this class the whitelist does not contain all the attributes.

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php [hunk 27]
```diff
@@ -173,4 +185,14 @@ public function setSensitiveData(SensitiveData $sensitiveData)
...
     }
+
+    public function obtainUserData(): array
+    {
+        $serializedPublicUserData = $this->serialize();
+        $serializedSensitiveUserData = $this->getSensitiveData()->serialize();
+        $serializedCombinedUserData = array_merge($serializedPublicUserData, $serializedSensitiveUserData);
+        $whitelist = array_flip(self::$whitelist);
+        return array_intersect_key($serializedCombinedUserData, $whitelist);
+//        return $serializedCombinedUserData;
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php:196]: A dev left over? Please remove the stray commented line of code!

File: src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php [hunk 28]
```diff
@@ -32,13 +32,27 @@
...
+        'identity_id',
+        'name_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+        'preferred_locale',
+        'email',
+        'common_name',
+        'second_factor_identifier',
+        'vettingType',
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/SecondFactorVettedEvent.php:53]: This seems to be incorrect. Looking at the Sensitive Data entity, this should be: `vetting_type`

File: src/Surfnet/Stepup/Identity/Event/WhitelistCreatedEvent.php [hunk 29]
```diff
@@ -19,9 +19,14 @@
 namespace Surfnet\Stepup\Identity\Event;
 
 use Surfnet\Stepup\Identity\Collection\InstitutionCollection;
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\RightToObtainDataInterface;
 
-class WhitelistCreatedEvent implements WhitelistEvent
+class WhitelistCreatedEvent implements WhitelistEvent, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'whitelisted_institutions',
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/WhitelistCreatedEvent.php:27]: @pmeulen Am I right in the assumtion that the events associated to the whitelist actions are be candidates to skip. They seem irrelevant to the Identity from a data portability perspective. But I might be wrong in that aspect? Whats your take on this?

File: src/Surfnet/Stepup/Identity/Event/SecondFactorVettedWithoutTokenProofOfPossession.php [hunk 30]
```diff
@@ -32,13 +32,27 @@
...
+        'identity_id',
+        'name_id',
+        'identity_institution',
+        'second_factor_id',
+        'second_factor_type',
+        'preferred_locale',
+        'email',
+        'common_name',
+        'second_factor_identifier',
+        'vettingType',
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/SecondFactorVettedWithoutTokenProofOfPossession.php:53]: Looking at SensitiveData serialize: this should be: `vetting_type`

**Review by MKodde — COMMENTED** [review]
Almost there @VadimSchmitz 

See some new feedback below. And maybe this is a good time for @pmeulen to join the review?

File: src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php (hunk 27 above)
Comment (VadimSchmitz) [reply src/Surfnet/Stepup/Identity/Event/PhonePossessionProvenEvent.php:196]: aremoved

File: src/Surfnet/StepupMiddleware/MiddlewareBundle/Service/EventStreamReplayer.php [hunk 31]
```diff
@@ -60,7 +60,6 @@ class EventStreamReplayer
         'audit_log',
         'ra_listing',
         'second_factor_revocation',
-        'whitelist_entry',
```
Comment (MKodde) [inline src/Surfnet/StepupMiddleware/MiddlewareBundle/Service/EventStreamReplayer.php:63]: Why did you remove the whitelist_entry? Also see line 72

File: src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaaForInstitutionEvent.php [hunk 32]
```diff
@@ -117,4 +128,9 @@ public function serialize(): array
             'ra_institution'              => (string) $this->raInstitution,
         ];
     }
+
+    public function obtainUserData(): array
+    {
+        // TODO: Implement obtainUserData() method.
+    }
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/IdentityAccreditedAsRaaForInstitutionEvent.php:135]: Forgot to implement this method?

File: src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedEvent.php [hunk 33]
```diff
@@ -25,13 +25,20 @@
...
 /**
  * @deprecated This event is superseded by the RegistrationAuthorityRetractedForInstitutionEvent because an RA institution was needed
  */
-class RegistrationAuthorityRetractedEvent extends IdentityEvent implements Forgettable
+class RegistrationAuthorityRetractedEvent extends IdentityEvent implements Forgettable, RightToObtainDataInterface
 {
+    protected static $whitelist = [
+        'identity_id',
+        'identity_institution',
+        'name_id',
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/RegistrationAuthorityRetractedEvent.php:39]: Sensitive data is not included in whitelist

File: src/Surfnet/Stepup/Identity/Event/SecondFactorMigratedEvent.php [hunk 34]
```diff
@@ -30,13 +30,25 @@
...
+    protected static $whitelist = [
+        'identity_id',
+        'source_institution',
+        'target_name_id',
+        'identity_institution',
+        'second_factor_id',
+        'new_second_factor_id',
+        'second_factor_type',
+        'preferred_locale',
+    ];
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Event/SecondFactorMigratedEvent.php:50]: Sensitive data is missing here? mail, commonname, sf identifier

**Review by MKodde — CHANGES_REQUESTED** [review]
For issues remain. Please, this time do not resolve the issues yourself. This makes it easier for me to check the progress. Thanks!

**Review by MKodde — APPROVED** [review]

File: src/Surfnet/StepupMiddleware/CommandHandlingBundle/SensitiveData/RightToObtainDataInterface.php [hunk 35]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData;
+
+interface RightToObtainDataInterface
```
Comment (pablothedude) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/SensitiveData/RightToObtainDataInterface.php:5]: Please consider to get only the field names with user data of the events with this interface and to do the aggregation of the data in a separate class to prevent the code duplication of merging data  in all concrete event classes.

File: src/Surfnet/Stepup/Identity/Event/WhitelistCreatedEvent.php (hunk 29 above)
Comment (pmeulen) [reply src/Surfnet/Stepup/Identity/Event/WhitelistCreatedEvent.php:27]: I agree. Events that handle middleware, institution and whitelist configuration do not contain any personal data and must be skipped.

File: src/Surfnet/StepupMiddleware/CommandHandlingBundle/SensitiveData/RightToObtainDataInterface.php (hunk 35 above)
Comment (MKodde) [reply src/Surfnet/StepupMiddleware/CommandHandlingBundle/SensitiveData/RightToObtainDataInterface.php:5]: Hi Bas, Thanks for the feedback. This is exactly what we'll do next. This was mainly an exercise for Vadim to get the whitelisted data right. The recurring filter logic will be offloaded to a helper class that is responsible for that task.

I'll be sure to include you on a successive PR where we implement this 'the right way'
